### ISSUE TITLE
How to 'unfold' children ?

### ISSUE THREAD

**Issue by gre** [facebook/react#4697]
Hi,

I'm writing a library which uses JSX as an abstract DSL, a bit like `react-router` have `<Router/>` and `<Route />`, or a bit like `react-canvas` have `<Surface />` and others primitives.

**To explain my question, I will take the example of `react-router`:**

An user could define a CustomRoute component that render() to a `<Route>`. 

and use it like this:

```
<Router>
  <CustomRoute>
</Router>
```

AFAIK This won't be detected by the `<Router/>` , because just traversing the props.children is not enough (router will see a `child.type==CustomRoute`, and won't understand it).

**Do you know if this is possible to solve this?**
Basically I want to descent the render() tree, and stop if it is not a user component (== it is a native component like `<div/>`), or if I reach the `<Route />` I'm looking for.

**Comment by jimfb** [comment]
When descending element.props.children, you can check `typeof element.type == 'function'` or `typeof element.type == 'string'`.  That will tell you if this is a custom component or a native component.  For custom components, you can (more or less) just say `var cls = element.type; var instance = new cls(); instance.props = element.props; var newElements = instance.render()` which will give you whatever that component would render to.  Obviously if the child components expect their lifecycles to be called, you'll need to call those, and similarly merge the component's default props if the child component expects those.  Getting all the detailed semantics right is slightly non-trivial but very doable.  If you're always trying to get to native elements, you might try using `renderToString` which will do all the heavy lifting for you.

For future reference, questions like this are better addressed on StackOverflow.  Github issues are reserved for tracking bugs in React.

**Comment by spicyj** [comment]
Eventually we may be able to support this by letting you implement a custom renderer. It's difficult right now, sorry.

**Comment by gre** [comment]
@spicyj at the time of my question I was heavily working on [gl-react](https://github.com/ProjectSeptemberInc/gl-react) but it was not yet public, that's why my question was a bit indirect 😄😄 

anyway, gl-react needs to traverse the user's VDOM to retrieve the `GL.Node` VDOM nodes. But more importantly, it deduplicates the nodes of the "GL graph" so only one OpenGL computation is done for shared nodes (explained in http://greweb.me/2016/06/glreactconf )

The way we current "unfold" the tree is by implementing what @jimfb said one year ago in the comment bellow. Thanks for this, I just notice I didn't replied at the time, it works fine :)
The only downside is we don't benefit at all the React lifecycle (it's fine for pure function components, and this is why GL Component can only be created with function at the moment). It's a bit limited to not be able to do "normal GL" on the rendering tree.
I'm wondering if other libraries like `react-art` do something better (like can you decorate a component with something like react-redux in the middle of the react-art tree?)

More generically, do React, since it separated to multiple implementations, now implements something to allow this use-case more easily or is it still a complex topic?

Cheers

**Comment by spicyj** [comment]
> like can you decorate a component with something like react-redux in the middle of the react-art tree

Yes. There is a restriction right now that you can't swap the type of element that a composite returns (ex: making render return `<A />` then return `<B />` after a state update breaks) but otherwise it works fully.

> More generically, do React, since it separated to multiple implementations, now implements something to allow this use-case more easily or is it still a complex topic? (like some new internal helpers provided?)

Still unsupported, sorry. I have not reviewed it but https://github.com/iamdustan/tiny-react-renderer claims to walk through building your own React rendering backend.

**Comment by gre** [comment]
thanks for the answer :)

**Comment by spicyj** [comment]
I should add: the react-art restriction is only because it embeds a new type inside a DOM tree and React doesn't really have the logic to support that. Standalone environments like DOM, Native, etc. don't have that problem.
